Scroll controls, expander, and refactor.

notlee opened the pull request:
Todo:

- [x] Add codedocs
- [x] Improve test coverage (public api, scroll header duplication)

### Improves Sort

Supports sort of more data types and custom data types.
Reviewed here: https://github.com/Financial-Times/o-table/pull/101

![sort](https://user-images.githubusercontent.com/10405691/45884740-9a90b700-bdac-11e8-8368-c55e374e4449.gif)


### Adds Expander & Scroll Controls

![expander-scroll](https://user-images.githubusercontent.com/10405691/45883392-e0e41700-bda8-11e8-931a-02ec1a4e19c7.gif)

### Improves Accessibility

Uses `<button>` with a title e.g. "sort table by Genus" to indicate the ability to sort (I could not use `aria-label`, as Voiceover would read the sort button label when moving between columns within the table body).

Duplicates headers for the "responsive scroll" table to keep the correct table roles and sensible keyboard navigation.

### Refactors JS

Makes more methods private and simplifies interfaces e.g. `sortRowsByColumn(columnIndex, sortOrder)` over `sortRowsByColumn(index, sortAscending, isNumericValue, type) `.

Splits the three responsive tables into their own classes to support unique functionality.

### Refactors SCSS

Consolidates utility mixins such as `oTableCellVerticallyCenter` with `oTableBase`.

Removes the ability to customise the base `o-table` class `$classname`. This means:
- More readable SCSS and simpler JS.
- Mixins can be used to granularly include `o-table` features without wrapping mixins in classes.
- Encourages including `o-table` once rather than multiple times with different class names.

Removes deprecated colour use cases.

Removes caption styles (except for alignment) in favour of placing a header within the caption.
<img width="772" alt="screen shot 2018-09-21 at 14 57 31" src="https://user-images.githubusercontent.com/10405691/45885611-cdd44580-bdae-11e8-89fb-deedb5522a46.png">


### Better Demos
🎉

gvonkoss commented:
🧀 😂

gvonkoss commented on src/scss/_base.scss:
```diff
@@ -1,80 +1,58 @@
-				vertical-align: top;
+
+		th {
+			@include oTypographySansBold(1);
```
[`oTypographyBold('sans')`](https://github.com/Financial-Times/o-typography/blob/079112ec83b024da2a22e714ee5dba0be871a427/src/scss/_mixins.scss#L149) doesn't rewrite the font-size or font-family

gvonkoss replied:
(But this only works if oTypographySans has already been applied to the element, and we're applying individually to `ths` and `tds` here?)

gvonkoss commented on src/scss/_base.scss:
```diff
@@ -1,80 +1,58 @@
-				vertical-align: top;
+
+		td {
+			@include oTypographySans(1);
```
☝️

gvonkoss commented on src/scss/_base.scss:
```diff
@@ -1,80 +1,58 @@
-	}
-}
+		.o-table__cell--content-secondary {
+			@include oTypographySans(0);
```
if a font was applied to the whole table, we could use [`oTypographySize(0)`](https://github.com/Financial-Times/o-typography/blob/079112ec83b024da2a22e714ee5dba0be871a427/src/scss/_mixins.scss#L71)

notlee replied:
Updated to `oTypographySize(0)` 👍

gvonkoss commented on src/scss/_borders.scss:
```diff
@@ -1,27 +1,25 @@
+		border-right: 1px solid _oTableGet('table-border-color');
+
+		// sass-lint:disable no-qualifying-elements
+		th[scope=row] {
```
if we don't enforce`[scope=row]`, we'll lose this styling, and its only in the markup, right?

notlee replied:
Correct, but we should enforce `scope=row`?

notlee replied:
👍

gvonkoss commented on src/scss/_compact.scss:
```diff
@@ -1,15 +1,10 @@
+	.o-table--compact {
+		th,
+		td {
+			@include oTypographySize(0);
```
same as other type comments ☝️

notlee replied:
I've updated this to:

```
		th {
			@include oTypographySansBold(0);
		}

		td {
			@include oTypographySans(0);
		}
```

So that progressively loaded fonts are handled correctly.

gvonkoss commented on src/scss/_mixins.scss:
```diff
@@ -1,80 +1,37 @@
 	}
 
+	// Stripes.
 	@if _oTableSupports('stripe') {
```
#überpedantic but stripe could be stripes if the mixin is stripes and a row has many stripes? 😬

gvonkoss commented on src/scss/_overlays.scss:
```diff
@@ -0,0 +1,152 @@
+	}
+
+	.o-table-fade-overlay--scroll {
+		background: linear-gradient(to right, _oTableGet('table-background') 2px, rgba(_oTableGet('table-background'), 0) var(--o-table-fade-from-start)),
```
💥

gvonkoss commented on src/scss/_sort.scss:
```diff
@@ -0,0 +1,48 @@
+		}
+	}
+
+	// Show ascending icon in sort button with ascending sort applied or on hover with no sort.
```
Maybe a [Matt Hinchliffe maneuvre](https://github.com/Financial-Times/o-expander/blob/6af37b58bed227211362fe1cef603278dc869f15/main.scss#L40)?

notlee replied:
Haha good shout. Um. My preference is to stick to the correct icon just in case they change and spinning it doesn't work. I confess that feels unlikely.

gvonkoss commented on README.md:
```diff
@@ -43,176 +29,154 @@ Where table headings (`th`) are used as row headings, `scope="row"` attributes m
-On a narrow viewpoint the Responsive Scroll version will move the headers to the right hand side of the table, and be fixed. This allows the data to be displayed in a column format which would allow the user to swipe left or right going through the data.
+#### Silent mode
 
-##### Overflow
```
(may be includes - the second includes in the sentences)

gvonkoss commented on README.md:
````diff
@@ -43,176 +29,154 @@ Where table headings (`th`) are used as row headings, `scope="row"` attributes m
+Required:
+```scss
+// Basic styles and utilities.
+@include oTableBase;
````
I mas have misread your explanation in the readme about the granularity, but I'm not sure I understand 100%

notlee replied:
The ability to use custom classes has been removed (reasoning in the PR description). I was trying to express that to include `o-table` features granularly.

Do this:
```scss
@include oTableBase;
@include oTableWrapper;
@include oTableContainer;
@include oTableResponsiveFlat;
...
```

Instead of this:
```scss
.o-table {
    @include oTableBase;
    @include oTableResponsiveFlat;
}
.o-table-wrapper {
    @include oTableWrapper;
}
.o-table-container {
    @include oTableContainer;
}
...
```

I only spoke to Jake and a few others casually about this, let me know if you have any questions or objections.

gvonkoss commented on README.md:
```diff
@@ -437,12 +414,52 @@ document.addEventListener('oTable.sorting', (event) => {
+	- `oTableAll`: does not accept a class name `$classname`. Instead use the default `o-table` class name -- please [contact us](#contact) if this is a problem for your team.
+	- `oTableCaptionTop`, `oTableCaptionBottom` have been removed. Use only `oTableCaption` instead.
+	- `oTableCellNumeric`, `oTableCellContentSecondary`, `oTableCellVerticallyCenter`, `oTableCaption` have been removed. Their styles are now included as part of `oTableBase`.
+	- All mixins now output classes directly, so must not be wrapped in an `.o-table` class manually.
```
this again relates to the explanation in the description of the PR. So scrap that.

gvonkoss replied:
No I still think points 1 and 4 are related

notlee replied:
I'm not sure I follow? `oTableAll` also.

notlee replied:
I've added a diff to that part of the migration guide which I hope will make things clearer. Let me know if now.

gvonkoss replied:
👌

notlee commented on README.md:
```diff
@@ -43,176 +29,154 @@ Where table headings (`th`) are used as row headings, `scope="row"` attributes m
-Instantiating an o-table JS component will add click events on the columns to trigger sorting.
-
-If using OBT:
+Instantiate `o-table` using Bower:
```
Hmm perhaps I shouldn't be mentioning Bower specifically here. I meant "not via the build service".

notlee commented on bower.json:
```diff
@@ -1,21 +1,23 @@
+    "o-grid": "^4.0.6",
+    "o-icons": ">=4.0.0 <6",
+    "o-typography": "^5.0.0",
+    "o-brand": ">=2.2.0 <4",
```
This is so we could update the major of `o-brand` one component at a time without releasing a breaking. I can set this to 3 now :)

notlee commented on src/js/Tables/BaseTable.js:
```diff
@@ -0,0 +1,137 @@
+			delete this.wrapper;
+			delete this.container;
+		} catch (error) {
+			// Elements never existed or are already deleted.
```
It's meant to fail silently, it's it's removed then that isn't an issue.

notlee replied:
Note: I haven't seen this error in `o-table` but have in `o-overlay`. It might be overly defensive.

notlee commented on src/js/Tables/OverflowTable.js:
```diff
@@ -0,0 +1,469 @@
+		}
+
+		// Add other event listeners to update controls.
+		this.wrapper.addEventListener('scroll', updateScoll);
```
I had to hit `cmd+f` and search `updateScroll` to prove that there was a difference. Legitimately could not see the missing `r`. 🙈

notlee commented on src/js/Tables/OverflowTable.js:
```diff
@@ -0,0 +1,469 @@
+				this.controls.controlsOverlay.style.display = '';
+				this.controls.fadeOverlay.style.display = '';
+				this._controlUpdateScheduled = false;
+			}.bind(this), 33);
```
Quite arbitrary. Roughly 30fps. This could be better. 😞I didn't want to update more frequently as it's only relevant as you reach the start/end of the table. I could not figure out a way to use the intersection observer in this case, as the controls are a sticky child of the container and not of the scrollable wrapper.

notlee commented:
@gvonkoss I think I've addressed all your feedback, either in code or a reply. Let me know if there's anything I missed or should revisit.

Since your review I've updated tests and made the following change as discussed:
- Controls wrapper, scroll buttons, and expander button is now specific to the overflow style of responsive table: that means `@include oTableControlsOverlay; @include oTableFadeOverlay;` is replaced with `@include oTableResponsiveOverflow;`.

Also also:
- `addSortButtons` method is now public.

gvonkoss commented on src/js/Tables/FlatTable.js:
```diff
@@ -0,0 +1,46 @@
+		if (this.tableHeaders.length > 0) {
+			this._duplicateHeaders(rootEl);
+		} else {
+			console.warn('Could create a "flat" table as no headers were found. Ensure table headers are placed within "<thead>". Removing class "o-table--responsive-flat".', rootEl);
```
Could * not * ?

notlee replied:
😬

gvonkoss commented on src/scss/_row-headings.scss:
```diff
@@ -1,12 +1,9 @@
-	&>tbody th {
-		background-color: _oTableGet('table-alternate-background');
+	.o-table--row-headings {
+		&>tbody th {
```
Do we need the & here? 🤔

notlee replied:
We do not 👍

notlee commented on src/scss/_sort.scss:
```diff
@@ -37,8 +37,8 @@
-	[aria-sort='none'] .o-table__sort:hover,
-	:not([aria-sort]) .o-table__sort:hover,
+	th[aria-sort='none']:hover .o-table__sort, // sass-lint:disable-line no-qualifying-elements
+	th:not([aria-sort]):hover .o-table__sort, // sass-lint:disable-line no-qualifying-elements
```
It's not that they're non-sortable, it's that they haven't been sorted yet (default sort is ascending).

notlee commented:
@gvonkoss 

>We mentioned that the class thing is tbd based on the proposal, but also, do we have a version of the table that gives the row headings some kind of prominence/background?

No, other than the boldness. 😞I think that would be a good addition.
